iroh-willow/src/store.rs

```diff
@@ -0,0 +1,144 @@
+    form::{AuthForm, EntryForm, EntryOrForm, SubspaceForm, TimestampForm},
+    interest::{CapSelector, UserSelector},
+    proto::{
+        data_model::Entry,
+        data_model::{AuthorisedEntry, PayloadDigest},
+        keys::{NamespaceId, NamespaceKind, NamespaceSecretKey, UserId},
+    },
+    store::traits::SecretStorage,
+    util::time::system_time_now,
+};
+
+use self::auth::{Auth, AuthError};
+use self::traits::Storage;
+
+pub(crate) use self::traits::EntryOrigin;
+
+pub(crate) mod auth;
+pub mod memory;
+pub mod traits;
+
+/// Storage for the Willow engine.
+///
+/// Wraps a `Storage` instance and adds the [`Auth`] struct that uses the secret and caps storage to provide
+/// authentication when inserting entries.
+#[derive(Debug, Clone)]
+pub(crate) struct Store<S: Storage> {
+    storage: S,
+    auth: Auth<S>,
+}
+
+impl<S: Storage> Store<S> {
+    pub fn new(storage: S) -> Self {
+        Self {
+            auth: Auth::new(storage.secrets().clone(), storage.caps().clone()),
+            storage,
+        }
+    }
+
+    pub fn entries(&self) -> &S::Entries {
+        self.storage.entries()
+    }
+
+    pub fn secrets(&self) -> &S::Secrets {
+        self.storage.secrets()
+    }
+
+    pub fn payloads(&self) -> &S::Payloads {
+        self.storage.payloads()
+    }
+
+    pub fn auth(&self) -> &Auth<S> {
+        &self.auth
+    }
+
+    pub async fn insert_entry(
+        &self,
+        entry: EntryOrForm,
+        auth: AuthForm,
+    ) -> Result<(AuthorisedEntry, bool)> {
+        let user_id = auth.user_id();
+        let entry = match entry {
+            EntryOrForm::Entry(entry) => Ok(entry),
+            EntryOrForm::Form(form) => self.form_to_entry(form, user_id).await,
+        }?;
+        let capability = match auth {
+            AuthForm::Exact(cap) => cap,
+            AuthForm::Any(user_id) => {
+                let selector = CapSelector::for_entry(&entry, UserSelector::Exact(user_id));
+                self.auth()
+                    .get_write_cap(&selector)?
+                    .ok_or_else(|| anyhow!("no write capability available"))?
+            }
+        };
+        let secret_key = self
+            .secrets()
+            .get_user(&user_id)
+            .context("Missing user keypair")?;
+
+        // TODO(frando): This should use `authorisation_token_unchecked` if we uphold the invariant
+        // that `user_id` is a pubkey for `secret_key`. However, that is `unsafe` at the moment
+        // (but should not be, IMO).
+        // Not using the `_unchecked` variant has the cost of an additional signature verification,
+        // so significant.
+        let token = capability.authorisation_token(&entry, secret_key)?;
+        let authorised_entry = AuthorisedEntry::new_unchecked(entry, token);
+        let inserted = self
+            .entries()
+            .ingest_entry(&authorised_entry, EntryOrigin::Local)?;
+        Ok((authorised_entry, inserted))
+    }
+
+    pub fn create_namespace(
```

Comment:
Working with namespaces having only their IDs is not very convenient. It would be nice to have some kind of association of namespace ID with some developer provided (in form of opaque data container) readable name like enum variant or constant string. Otherwise, one have to create its own association map with some persistence logic, partially duplicating namespace store as result of that.